akka-stream/src/main/scala/akka/stream/impl/Throttle.scala

```diff
@@ -20,21 +35,32 @@ private[stream] class Throttle[T](cost: Int,
                                   mode: ThrottleMode)
   extends SimpleLinearGraphStage[T] {
   require(cost > 0, "cost must be > 0")
-  require(per.toMillis > 0, "per time must be > 0")
+  require(per.toNanos > 0, "per time must be > 0")
   require(!(mode == ThrottleMode.Enforcing && maximumBurst < 0), "maximumBurst must be > 0 in Enforcing mode")
+  require(per.toNanos >= cost, "Rates larger than 1 unit / nanosecond are not supported")
+
+  // There is some loss of precision here because of rounding, but this only happens if nanosBetweenTokens is very
+  // small which is usually at rates where that precision is highly unlikely anyway as the overhead of this stage
+  // is likely higher than the required accuracy interval.
+  private val nanosBetweenTokens = (per.toNanos.toDouble / cost).toLong
+  private val timerName: String = "ThrottleTimer"
 
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
     var willStop = false
-    var lastTokens: Long = maximumBurst
-    var previousTime: Long = now()
-
-    val speed = ((cost.toDouble / per.toNanos) * 1073741824).toLong
-    val timerName: String = "ThrottleTimer"
+    var recordedTokens: Long = maximumBurst
+    // Tracks when the last registered token arrived, i.e. the ideal time when the *tracked* bucket jumped
+    // from n to n+1 tokens (but not when it was actually calculated).
+    // This needs to be only updated for token *additions* as subtractions have nothing to do with token arrival.
+    // If 5 tokens arrived while 4 were consumed, this needs to be still increased by 5 * nanosBetweenTokens.
+    // Be aware that this timestamp is usually in the past compared to when the ideal bucket had a new token arrival.
+    // We track here only the timestamp of the last *registered* token, i.e. corresponding to the last addition to
+    // recordedTokens (this is due to our on-need, lazy bucket update policy which makes us always slightly be in
+    // the past compared to the ideal bucket, doing large jumps in the future only when update is needed).
+    var arrivalOfLastRecordedToken: Long = System.nanoTime()
```

Comment:
this could be 0 actually, since we update it in preStart() (which is slightly closer in time to the "real" start).